### PR TITLE
feat(xds): Add OVHcloud service discovery to the ScrapeConfig CRD

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

support OVHCloud xds
see: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add OVHcloud service discovery to the ScrapeConfig CRD
```

### CODE REVIEW

**Comment by slashpai** [comment]
@dongjiang1989 Currently we are still reviewing what all SDs we want to support you can take a look at the design proposal https://github.com/prometheus-operator/prometheus-operator/pull/6634

**Comment by dongjiang1989** [comment]
> @dongjiang1989 Currently we are still reviewing what all SDs we want to support you can take a look at the design proposal #6634

Got it. thanks

**Comment by slashpai** [comment]
@mviswanathsai can you review? :)

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -4026,6 +4026,56 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+	// OVHCloudSDConfigs
+	if len(sc.Spec.OVHCloudSDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.OVHCloudSDConfigs))
+		for i, config := range sc.Spec.OVHCloudSDConfigs {
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "application_key",
+				Value: config.ApplicationKey,
+			})
+
+			value, _ := store.GetSecretKey(ctx, sc.GetNamespace(), *config.ApplicationSecret)
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4038]: Wouldn't we want to handle an error here, in case there is one? 🤔

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -4026,6 +4026,56 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "application_secret",
+				Value: value,
+			})
+
+			key, _ := store.GetSecretKey(ctx, sc.GetNamespace(), *config.ConsumerKey)
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4044]: Same remark here.

File: pkg/prometheus/promcfg_test.go [hunk 3]
```diff
@@ -9776,6 +9776,89 @@ func TestScrapeConfigSpecConfigWithLightSailSD(t *testing.T) {
 		})
 	}
 }
+
+func TestScrapeConfigSpecConfigWithOVHCloudSD(t *testing.T) {
```
Comment (mviswanathsai) [inline pkg/prometheus/promcfg_test.go:9780]: (nit) could you also add the `Service` field to the test case?

File: pkg/prometheus/resource_selector_test.go [hunk 4]
```diff
@@ -3393,6 +3393,29 @@ func TestSelectScrapeConfigs(t *testing.T) {
 			},
 			selected: false,
 		},
+		{
+			scenario: "OVHCloud SD config",
```
Comment (mviswanathsai) [inline pkg/prometheus/resource_selector_test.go:3397]: It need not be done in this same PR, but tests with faulty (non-existent) SecretKeySelectors might also be a good addition.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -939,6 +942,31 @@ type NomadSDConfig struct {
...
+type OVHCloudSDConfig struct {
+	// Access key to use. https://api.ovh.com.
+	// +required
+	ApplicationKey string `json:"applicationKey"`
+	// +required
+	ApplicationSecret *corev1.SecretKeySelector `json:"applicationSecret"`
+	// +required
+	ConsumerKey *corev1.SecretKeySelector `json:"consumerKey"`
+	// Service of the targets to retrieve. Must be `vps` or `dedicated_server`.
+	// +kubebuilder:validation:Enum=vps;dedicated_server
```
[the remark this answers is not in the document]
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: We may want to allow both lowercase and uppercase here but always convert in config.

[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: I'd stick with only 1 value per enum and pick `DedicatedServer` to follow https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#constants

[the remark this answers is not in the document]
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: I think we had a similar request for allowing both it was Role field. IIRC we added both later. 

May be we want to allow it only based on request or we want to be restrictive?

File: pkg/prometheus/promcfg.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4044]: Same

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -4026,6 +4026,56 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "consumer_key",
+				Value: key,
+			})
+
+			if config.Service != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
+					Key:   "service",
+					Value: config.Service,
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4053]: Thanks. 
Is it more reasonable to maintain the user-defined capitalization specifications of CRD? 🤔️

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: ref: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config

```
# Service of the targets to retrieve. Must be `vps` or `dedicated_server`.
service: <string>
```

Allow users to customize the case, and the operator will not do excessive escaping.

File: pkg/prometheus/resource_selector_test.go (hunk 4 above)
Comment (dongjiang1989) [reply pkg/prometheus/resource_selector_test.go:3397]: Got it.

File: pkg/prometheus/promcfg_test.go (hunk 3 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg_test.go:9780]: Got it. Done.

File: pkg/prometheus/promcfg.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/prometheus/promcfg.go:4053]: > Is it more reasonable to maintain the user-defined capitalization specifications of CRD? 🤔️

Not sure to follow here, could you elaborate?

[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:4053]: > Not sure to follow here, could you elaborate?

ref: 
https://github.com/prometheus/prometheus/blob/5c417684f87f2e7f55b172b3e5e5c96ac9bc2e2c/discovery/ovhcloud/ovhcloud.go#L89-L94

For me, remove the enum limit. It's best to keep it vague. 
/cc @simonpasquier @slashpai

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (mviswanathsai) [reply pkg/prometheus/promcfg.go:4038]: My bad. Since we have already added the secret key to store, we can just make use of the `cacheOnlyStore` and get the key from the cache like we do for [BasicAuth](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L438-L463), OAuth2 etc. 
```suggestion
			value, err := store.GetSecretKey(*config.ApplicationSecret)
			if err != nil {
			            level.Error(cg.logger).Log("err", fmt.Sprintf("invalid applicationSecret ref: %s", err))
		                    return cfg 
                         }
```

File: pkg/prometheus/promcfg.go (hunk 2 above)
Comment (mviswanathsai) [reply pkg/prometheus/promcfg.go:4044]: Applies here as well.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: > I'd stick with only 1 value per enum and pick `DedicatedServer` to follow https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#constants

Thanks.
If change `dedicated_server` to  `DedicatedServer`, Operator needs to maintain these mapping relationships. 
Is it necessary to maintain such a relationship? 🤔️ @simonpasquier

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 7]
```diff
@@ -939,6 +942,31 @@ type NomadSDConfig struct {
...
+
+// OVHCloudSDConfig configurations allow retrieving scrape targets from OVHcloud's dedicated servers and VPS using their API.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config
+// +k8s:openapi-gen=true
+type OVHCloudSDConfig struct {
+	// Access key to use. https://api.ovh.com.
+	// +required
+	ApplicationKey string `json:"applicationKey"`
+	// +required
+	ApplicationSecret *corev1.SecretKeySelector `json:"applicationSecret"`
```
Comment (mviswanathsai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:954]: Since this is a required field, can we just use the value here instead of a pointer?
```suggestion
	ApplicationSecret corev1.SecretKeySelector `json:"applicationSecret"`
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 8]
```diff
@@ -939,6 +942,31 @@ type NomadSDConfig struct {
...
+	// +required
+	ConsumerKey *corev1.SecretKeySelector `json:"consumerKey"`
```
Comment (mviswanathsai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:956]: Same remark here as well.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: My opinion is lets make it restrictive and add the other cases based on requests.

File: pkg/prometheus/promcfg.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/prometheus/promcfg.go:4053]: > It's best to keep it vague.

Not sure what you mean but my opinion is, if we can add valuable validation at the operator its best we do that. I'd go a step further and say its best if we can add most of the validations at the API level (i.e, make the API restrictive) and only touch the application level validation when absolutely necessary. e.g, when fetching secrets, etc.

TLDR; If Prometheus is validating for something (at the application level or otherwise) it is best that the operator do at least that. I think @simonpasquier and @slashpai will agree as well.

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4038]: My bad. Since we have already added the secret key to store, we can just make use of the cacheOnlyStore and get the key from the cache like we do for [BasicAuth](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L438-L463), OAuth2 etc.

```suggestion
			value, err := store.GetSecretKey(config.ApplicationSecret)
			if err != nil {
			            level.Error(cg.logger).Log("err", fmt.Sprintf("invalid applicationSecret ref: %s", err))
		                    return cfg 
                         }```

File: pkg/prometheus/promcfg.go (hunk 2 above)
Comment (mviswanathsai) [inline pkg/prometheus/promcfg.go:4044]: Same remark here as well.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: > If change dedicated_server to DedicatedServer, Operator needs to maintain these mapping relationships.
Is it necessary to maintain such a relationship?

Yes it means a bit of extra work for the operator but I'd prefer to follow the k8s conventions.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 9]
```diff
@@ -939,6 +942,31 @@ type NomadSDConfig struct {
...
+	// Service of the targets to retrieve. Must be `vps` or `dedicated_server`.
+	// +kubebuilder:validation:Enum=vps;dedicated_server
+	// +optional
+	Service *string `json:"service,omitempty"`
+	// Custom endpoint to be used.
+	// +kubebuilder:validation:MinLength=1
+	// +optional
+	Endpoint *string `json:"endpoint,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:964]: hmm looks like Endpoint is a required field.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 10]
```diff
@@ -939,6 +942,31 @@ type NomadSDConfig struct {
...
 	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
 }
+
+// OVHCloudSDConfig configurations allow retrieving scrape targets from OVHcloud's dedicated servers and VPS using their API.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config
+// +k8s:openapi-gen=true
+type OVHCloudSDConfig struct {
+	// Access key to use. https://api.ovh.com.
+	// +required
+	ApplicationKey string `json:"applicationKey"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:952]:
```suggestion
	// +kubebuilder:validation:MinLength=1
	ApplicationKey string `json:"applicationKey"`
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 9 above)
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:964]: hmm looks like Endpoint is a required field.

```suggestion
	// +required
	Endpoint string `json:"endpoint"`
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 7 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:954]: Got it.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 8 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:956]: Got it

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 9 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:964]: Thanks. My bad. Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 10 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:953]: Got it. Done

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 9 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Got it. Done

**Comment by dongjiang1989** [comment]
@simonpasquier @mviswanathsai  PTAL re-check

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 11]
```diff
@@ -939,6 +942,32 @@ type NomadSDConfig struct {
...
+	// +required
+	ApplicationSecret corev1.SecretKeySelector `json:"applicationSecret"`
+	// +required
+	ConsumerKey corev1.SecretKeySelector `json:"consumerKey"`
+	// Service of the targets to retrieve. Must be `VPS` or `DedicatedServer`.
+	// +kubebuilder:validation:Enum=VPS;DedicatedServer
+	// +optional
+	Service *string `json:"service,omitempty"`
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Got it.
Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 12]
```diff
@@ -939,6 +942,32 @@ type NomadSDConfig struct {
 	// +optional
 	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
 }
+
+// OVHCloudSDConfig configurations allow retrieving scrape targets from OVHcloud's dedicated servers and VPS using their API.
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ovhcloud_sd_config
+// +k8s:openapi-gen=true
+type OVHCloudSDConfig struct {
```
[the remark this answers is not in the document]
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: OVHCloudSDConfig currently does not support the common fields. 🤔️

File: pkg/prometheus/promcfg.go [hunk 13]
```diff
@@ -4024,6 +4024,58 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+	// OVHCloudSDConfigs
+	if len(sc.Spec.OVHCloudSDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.OVHCloudSDConfigs))
+		for i, config := range sc.Spec.OVHCloudSDConfigs {
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "application_key",
+				Value: config.ApplicationKey,
+			})
+
+			value, _ := s.GetSecretKey(config.ApplicationSecret)
```
[the remark this answers is not in the document]
Comment (mviswanathsai) [reply pkg/prometheus/promcfg.go:4004]: Got it!

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:958]: Good reason then :sweat_smile:

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 14]
```diff
@@ -940,6 +943,40 @@ type NomadSDConfig struct {
...
+	// +required
+	ApplicationKey string `json:"applicationKey"`
+	// +required
+	ApplicationSecret corev1.SecretKeySelector `json:"applicationSecret"`
+	// +required
+	ConsumerKey corev1.SecretKeySelector `json:"consumerKey"`
+	// Service of the targets to retrieve. Must be `VPS` or `DedicatedServer`.
+	// +kubebuilder:validation:Enum=VPS;DedicatedServer
+	// +optional
+	Service *OVHService `json:"service,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:970]: If I read the Prometheus code properly, service is mandatory field.
```suggestion
	Service OVHService `json:"service"`
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 15]
```diff
@@ -940,6 +943,40 @@ type NomadSDConfig struct {
...
+	// Custom endpoint to be used.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Endpoint string `json:"endpoint"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:974]: And conversely endpoint has a default value when unset so it should be optional but > 0 if set.
```suggestion
	// +optional
	Endpoint *string `json:"endpoint,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 14 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: My bad. Got it. 
Done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 15 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Thanks. Got it.
Done.

**Review by simonpasquier — APPROVED** [review]
Thanks!
